compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -9,14 +9,165 @@
 #include "iree-amd-aie/Transforms/Passes.h"
 #include "iree-amd-aie/aie_runtime/Utils/ChannelBdIdGenerator.h"
 #include "iree-amd-aie/aie_runtime/iree_aie_runtime.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
 
 #define DEBUG_TYPE "iree-amdaie-assign-npu-dma-bd-ids"
 
 namespace mlir::iree_compiler::AMDAIE {
 
 namespace {
 
-/// Assign BD ids to NPU dma operations using the BD generator
+// Utility to retrieve a TileOp from a vector of tile values, while doing
+// appropriate verifications.
+FailureOr<AMDAIE::TileOp> getGeneratorTileOp(
+    AMDAIE::NpuDmaCpyNdOp &npuDmaOp,
+    DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap) {
+  SmallVector<Value> tiles;
+  if (npuDmaOp.getSource()) {
+    auto logicalObjFifo =
+        dyn_cast_if_present<AMDAIE::LogicalObjFifoOpInterface>(
+            npuDmaOp.getSource().getDefiningOp());
+    if (!logicalObjFifo)
+      return npuDmaOp.emitOpError() << "expected a source logical objectFifo";
+    tiles = logicalObjFifo.getTiles();
+  }
+  if (npuDmaOp.getTarget()) {
```

Comment:
I think this doesn't result in exactly the same logic as before:

Before:

````
if (npuDmaOp.getSource()) {
  tiles = npuDmaOp.getSource()...getTiles();
  getGeneratorTileOp(tiles, ...);
}
if (npuDmaOp.getTarget()) {
  tiles = npuDmaOp.getSource()...getTiles();
  getGeneratorTileOp(tiles, ...);
}
```

Here, both source and target could be set and handled, while after this change:

````
getGeneratorTileOp() {
  tiles;
  if (npuDmaOp.getSource()) {
     tiles = npuDmaOp.getSource()...getTiles();
   }
   if (npuDmaOp.getTarget()) {
      tiles = npuDmaOp.getTarget()...getTiles();
   }
}
```

So, if target is set, tiles will always be overridden with the target tiles, so a potential `NpuDmaCpyNdOp` with two sides isn't supported anymore (we will need that in the future).

You could solve this for example by using `template <CopyOpOperateOn OperateOn>` on this function.